move/move-ibc/sources/ibc.move

```diff
@@ -1230,120 +928,481 @@ module IBC::Core {
+        let connection = smart_table::borrow(
+            &store.connections,
+            connection_id,
         );
+        if (connection_end::state(connection) != CONN_STATE_OPEN) {
+            abort(E_INVALID_CONNECTION_STATE)
+        };
+        *connection
+    }
 
-        table::upsert(
-            &mut store.commitments,
-            IBCCommitment::next_sequence_ack_commitment_key(port_id, channel_id),
-            bcs::to_bytes(&1)
+    // TODO(aeryz): borrow instead of copy
+    public fun ensure_connection_feature(connection_hops: vector<String>, ordering: u8): (String, ConnectionEnd) acquires IBCStore {
+        if (vector::length(&connection_hops) != 1) {
+            abort(E_CONN_NOT_SINGLE_HOP)
+        };
+        let connection_id = *vector::borrow(&connection_hops, 0);
+        let connection = ensure_connection_state(connection_id);
+        if (vector::length(connection_end::versions(&connection)) != 1) {
+            abort(E_CONN_NOT_SINGLE_VERSION)
+        };
+        let version = *vector::borrow(connection_end::versions(&connection), 0);
+        if (!verify_supported_feature(&version, to_string(ordering))) {
+            abort(E_UNSUPPORTED_FEATURE)
+        };
+        (connection_id, connection)
+    }
+
+    // TODO: Here the implementation returns false if there is any uppercase character in the string
+    // eg: "portId" will return false but "port-0" will return true. Is this correct way to check?
+    public fun is_lowercase(s: &String): bool {
+        let bytes = bcs::to_bytes(s);
+        let i = 1; // First byte is the length of the string
+        while (i < vector::length(&bytes)) {
+            let byte = *vector::borrow(&bytes, i);
+            // if (byte < 0x61 || byte > 0x7A) { // ASCII values for 'a' and 'z'
+            //     return false;
+            // };
+            if (byte > 0x40 && byte < 0x5b) {
+                return false
+            };
+            i = i + 1;
+        };
+        true
+    }
+    
+    public fun update_channel_commitment(port_id: String, channel_id: String) acquires IBCStore {
+        let store = borrow_global_mut<IBCStore>(get_vault_addr());
+        let channel_port = ChannelPort{port_id:port_id, channel_id};
+        let channel = smart_table::borrow(
+            &store.channels,
+            channel_port
         );
 
-        update_channel_commitment(port_id, channel_id);
+        let encoded = channel::encode_proto(*channel);
+        let key = IBCCommitment::channel_commitment_key(port_id, channel_id);
+        table::upsert(&mut store.commitments, key, hash::sha2_256(encoded));
+    }
 
-        claim_capability(
-            IBCCommitment::channel_capability_path(port_id, channel_id),
-            signer::address_of(ibc_app),
+
+    public fun verify_channel_state(
+        connection: &ConnectionEnd,
+        height: height::Height,
+        proof: vector<u8>,
+        port_id: String,
+        channel_id: String,
```

Comment:
does move have any concept of "zero cost newtypes"? might be nice to use something like that for the various id types instead of string directly.